File: Pattern.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace FileSquid
{
    /// <summary>
    /// Describes a general form of an object using placeholder variables.
    /// </summary>
    /// <typeparam name="K">An identifier for a variable.</typeparam>
    /// <typeparam name="T">The concrete object type this pattern is for (e.g. string).</typeparam>
    /// <typeparam name="M">The variable mapping type used by this pattern.</typeparam>
    public abstract class Pattern<K, T, M>
        where M : IMap<K, object, M>
    {
        /// <summary>
        /// Creates a variable pattern.
        /// </summary>
        public static Pattern<K, T, M> Variable(K Identifier)
        {
            return new Variable<K, T, M>(Identifier);
        }

        /// <summary>
        /// Creates a literal pattern.
        /// </summary>
        public static Pattern<K, T, M> Literal(T Value)
        {
            return new Literal<K, T, M>(Value);
        }

        /// <summary>
        /// Tries evaluating this pattern for the given variable values.
        /// </summary>
        public abstract bool Evaluate(M Variables, out T Value);

        /// <summary>
        /// Finds mappings of variables or values that would make this pattern equivalent to the given
        /// value.
        /// </summary>
        public abstract IEnumerable<M> Match(M Initial, T Value);

        /// <summary>
        /// Substitutes the given variable with the given pattern. Returns null if the type of the variable
        /// does not correspond with the type of the target pattern.
        /// </summary>
        public abstract Pattern<K, T, M> Substitute<NT>(K Variable, Pattern<K, NT, M> Pattern);

        /// <summary>
        /// Gets the domain (referenced variables along with their associated type) for this pattern.
        /// </summary>
        public Dictionary<K, Type> Domain
        {
            get
            {
                Dictionary<K, Type> domain = new Dictionary<K, Type>();
                this.WriteDomain(domain);
                return domain;
            }
        }

        /// <summary>
        /// Writes the domain of this pattern to the given set.
        /// </summary>
        public abstract void WriteDomain(Dictionary<K, Type> Domain);
    }

    /// <summary>
    /// Contains useful functions for creating and manipulating patterns.
    /// </summary>
    public static class Pattern
    {
        /// <summary>
        /// Concatenates the given string patterns.
        /// </summary>
        public static Pattern<K, string, M> Concat<K, M>(IEnumerable<Pattern<K, string, M>> Parts)
            where M : IMap<K, object, M>
        {
            return ConcatPattern<K, M>.Create(Parts);
        }
    }

    /// <summary>
    /// A pattern for an arbitrary variable.
    /// </summary>
    public sealed class Variable<K, T, M> : Pattern<K, T, M>
        where M : IMap<K, object, M>
    {
        public Variable(K Identifier)
        {
            this.Identifier = Identifier;
        }

        /// <summary>
        /// The identifier for this variable pattern.
        /// </summary>
        public readonly K Identifier;

        public override bool Evaluate(M Variables, out T Value)
        {
            object lvalue = Variables[this.Identifier];
            if (lvalue != null && lvalue is T)
            {
                Value = (T)lvalue;
                return true;
            }
            Value = default(T);
            return false;
        }

        public override IEnumerable<M> Match(M Initial, T Value)
        {
            object cur = Initial[this.Identifier];
            if (cur == null)
            {
                Initial[this.Identifier] = Value;
                return new M[] { Initial };
            }
            if (Value.Equals(cur))
            {
                return new M[] { Initial };
            }
            return new M[0];
        }

        public override Pattern<K, T, M> Substitute<NT>(K Variable, Pattern<K, NT, M> Pattern)
        {
            if (this.Identifier.Equals(Variable))
                return Pattern as Pattern<K, T, M>;
            else
                return this;
        }

        public override void WriteDomain(Dictionary<K, Type> Domain)
        {
            Domain[this.Identifier] = typeof(T);
        }
    }

    /// <summary>
    /// A pattern for a specific value.
    /// </summary>
    public sealed class Literal<K, T, M> : Pattern<K, T, M>
        where M : IMap<K, object, M>
    {
        public Literal(T Value)
        {
            this.Value = Value;
        }

        /// <summary>
        /// The value of this literal.
        /// </summary>
        public readonly T Value;

        public override bool Evaluate(M Variables, out T Value)
        {
            Value = this.Value;
            return true;
        }

        public override IEnumerable<M> Match(M Initial, T Value)
        {
            if (Value.Equals(this.Value))
                return new M[] { Initial };
            else
                return new M[0];
        }

        public override Pattern<K, T, M> Substitute<NT>(K Variable, Pattern<K, NT, M> Pattern)
        {
            return this;
        }

        public override void WriteDomain(Dictionary<K, Type> Domain)
        {

        }
    }

    /// <summary>
    /// A string pattern created by concatenating sub-patterns.
    /// </summary>
    public sealed class ConcatPattern<K, M> : Pattern<K, string, M>
        where M : IMap<K, object, M>
    {
        private ConcatPattern(Pattern<K, string, M>[] Parts)
        {
            this.Parts = Parts;
        }

        /// <summary>
        /// Creates a pattern by concatenating the given parts.
        /// </summary>
        public static Pattern<K, string, M> Create(IEnumerable<Pattern<K, string, M>> Parts)
        {
            List<Pattern<K, string, M>> nParts = new List<Pattern<K, string, M>>();
            string literalString = "";
            foreach (Pattern<K, string, M> part in Parts)
            {
                Literal<K, string, M> literal = part as Literal<K, string, M>;
                if (literal != null) 
                    literalString += literal.Value;
                else
                {
                    if (literalString != "")
                    {
                        nParts.Add(new Literal<K, string, M>(literalString));
                        literalString = "";
                    }
                    nParts.Add(part);
                }
            }
            if (nParts.Count == 0)
                return new Literal<K, string, M>(literalString);
            else
            {
                if (literalString != "")
                {
                    nParts.Add(new Literal<K, string, M>(literalString));
                    return new ConcatPattern<K, M>(nParts.ToArray());
                }
                else if (nParts.Count == 1) return nParts[0];
                else return new ConcatPattern<K, M>(nParts.ToArray());
            }
        }

        /// <summary>
        /// The parts of this pattern.
        /// </summary>
        public readonly Pattern<K, string, M>[] Parts;

        public override bool Evaluate(M Variables, out string Value)
        {
            Value = "";
            StringBuilder builder = new StringBuilder();
            for (int t = 0; t < this.Parts.Length; t++)
            {
                string str;
                if (!this.Parts[t].Evaluate(Variables, out str))
                    return false;
                builder.Append(str);
            }
            Value = builder.ToString();
            return true;
        }

        /// <summary>
        /// Determines whether two substrings are the same (using culture-insensitive equality checking).
        /// </summary>
        private static bool _SubstringEqual(string A, int AIndex, string B, int BIndex, int Length)
        {
            while (Length > 0)
            {
                if (A[AIndex] != B[BIndex]) return false;
                AIndex++;
                BIndex++;
                Length--;
            }
            return true;
        }

        /// <summary>
        /// Finds the next occurence of the given string at the given index (using culture-insensitive equality checking),
        /// returning -1 if it does not occur.
        /// </summary>
        private static int _NextOccurence(string Target, string String, int Index)
        {
            while (Index < String.Length - Target.Length)
            {
                for (int t = 0; t < Target.Length; t++)
                {
                    if (Target[t] != String[t + Index])
                        goto nomatch;
                }
                return Index;
            nomatch:
                Index++;
            }
            return -1;
        }

        /// <summary>
        /// An intermediate match result that takes into account matches of the leftmost
        /// parts of a concatenated pattern.
        /// </summary>
        private struct _PossibleMatch
        {
            /// <summary>
            /// The variable map for this match.
            /// </summary>
            public M Map;

            /// <summary>
            /// The index in the input string this possible match is on.
            /// </summary>
            public int Index;
        }

        public override IEnumerable<M> Match(M Initial, string Value)
        {
            int length = Value.Length;
            int partCount = this.Parts.Length;
            List<_PossibleMatch> possibleMatches = new List<_PossibleMatch>();

            // Match the prefix, if one exists.
            int nextPartIndex;
            Pattern<K, string, M> first = this.Parts[0];
            Literal<K, string, M> literal = first as Literal<K, string, M>;
            if (literal != null)
            {
                if (literal.Value.Length <= length && _SubstringEqual(literal.Value, 0, Value, 0, literal.Value.Length))
                {
                    first = this.Parts[1];
                    nextPartIndex = 2;
                    possibleMatches.Add(new _PossibleMatch { Map = Initial, Index = literal.Value.Length });
                }
                else
                {
                    return new M[0];
                }
            }
            else
            {
                nextPartIndex = 1;
                possibleMatches.Add(new _PossibleMatch { Map = Initial, Index = 0 });
            }

            // Match the suffix, if one exists.
            literal = this.Parts[partCount - 1] as Literal<K, string, M>;
            if (literal != null)
            {
                if (literal.Value.Length <= length && _SubstringEqual(literal.Value, 0, Value, length - literal.Value.Length, literal.Value.Length))
                {
                    length -= literal.Value.Length;
                    partCount--;
                }
                else
                {
                    return new M[0];
                }
            }


            // Match middle parts (after the prefix before the final part).
            Pattern<K, string, M> current = first;
            List<_PossibleMatch> nextMatches = new List<_PossibleMatch>();
            while (nextPartIndex < partCount)
            {
                if (possibleMatches.Count == 0) return new M[0];

                Pattern<K, string, M> next = this.Parts[nextPartIndex];
                if ((literal = next as Literal<K, string, M>) != null)
                {
                    foreach (_PossibleMatch p in possibleMatches)
                        for (int lstart = _NextOccurence(literal.Value, Value, p.Index); lstart != -1;
                            lstart = _NextOccurence(literal.Value, Value, lstart + 1))
                        {

                            foreach (M m in current.Match(p.Map.Copy(), Value.Substring(p.Index, lstart - p.Index)))
                                nextMatches.Add(new _PossibleMatch { Map = m, Index = lstart + literal.Value.Length });
                        }
                    nextPartIndex++;
                    next = this.Parts[nextPartIndex];
                }
                else
                {
                    foreach (_PossibleMatch p in possibleMatches)
                        for (int pend = p.Index; pend <= Value.Length; pend++)
                            foreach (M m in current.Match(p.Map.Copy (), Value.Substring(p.Index, pend - p.Index)))
                                nextMatches.Add(new _PossibleMatch { Map = m, Index = pend });
                }

                List<_PossibleMatch> temp = possibleMatches;
                possibleMatches = nextMatches;
                nextMatches = temp;
                nextMatches.Clear();
                current = next;
                nextPartIndex++;
            }

            // Perform final matches.
            List<M> results = new List<M>();
            foreach (_PossibleMatch p in possibleMatches)
                results.AddRange(current.Match(p.Map, Value.Substring(p.Index, length - p.Index)));
            return results;
        }

        public override Pattern<K, string, M> Substitute<NT>(K Variable, Pattern<K, NT, M> Pattern)
        {
            Pattern<K, string, M>[] nParts = new Pattern<K, string, M>[this.Parts.Length];
            for (int t = 0; t < nParts.Length; t++)
                nParts[t] = this.Parts[t].Substitute(Variable, Pattern);
            return Create(nParts);
        }

        public override void WriteDomain(Dictionary<K, Type> Domain)
        {
            foreach (Pattern<K, string, M> part in this.Parts)
                part.WriteDomain(Domain);
        }
    }
}

File: Pattern.fs
namespace FileSquid

open System
open System.Collections.Generic
open System.Text
open FileSquid.Util

/// Describes the general form of a value.
type Pattern =
    interface
    end

/// Describes the general form of a value of a certain type.
type [<AbstractClass>] Pattern<'a, 'b when 'b : equality> () =
    interface Pattern
    
    /// Evaluates this pattern for a given parameter, returning None if any of the variables
    /// are not set or set to an invalid value.
    abstract Evaluate : 'a -> 'b option

    /// Finds parameters for this pattern that would make it equivalent to the given target value, 
    /// appending them to the given list.
    abstract Match : 'b -> 'a list -> 'a  -> 'a list

    /// Substitutes the variables in this pattern using the given mapping function.
    abstract Substitute : (Variable -> Pattern) -> Pattern<'a, 'b>

/// A reference to a variable value within a parameter structure. 
and Variable =
    interface
    end

/// A reference to a typed variable value within a parameter structure.
type [<AbstractClass>] Variable<'a, 'b when 'b : equality> () =
    inherit Pattern<'a, 'b> ()
    interface Variable
    override this.Substitute map = (map this) :?> Pattern<'a, 'b>

/// A pattern with a specific value.
type Literal<'a, 'b when 'b : equality> (value : 'b) =
    inherit Pattern<'a, 'b> ()

    /// The value for this literal.
    member this.Value = value

    override this.Evaluate _ = Some value
    override this.Match target accum initial = 
        if value = target then initial :: accum
        else accum
    override this.Substitute _ = this :> Pattern<'a, 'b>

/// A pattern for a concatenation of strings.
type Concat<'a> private (parts : Pattern<'a, string> list) =
    inherit Pattern<'a, string> ()
    let (|Known|_|) initial (pattern : Pattern<'a, string>) = pattern.Evaluate initial

    /// Creates a concatenation pattern for the given parts.
    static member Create parts =
        let rec concat (literalString : string) accum parts =
            let forceAccum () = 
                if literalString.Length > 0 then (Literal<'a, string> literalString :> Pattern<'a, string>) :: accum
                else accum
            match (parts : Pattern<'a, string> list) with
            | :? Literal<'a, string> as part :: parts -> concat (literalString + part.Value) accum parts
            | :? Concat<'a> as part :: parts -> concat literalString accum (List.append part.Parts parts)
            | part :: parts -> concat "" (part :: forceAccum ()) parts
            | [] -> 
                match forceAccum () with
                | [] -> Literal<'a, string> "" :> Pattern<'a, string>
                | accum -> Concat<'a> (List.rev accum) :> Pattern<'a, string>
        concat "" [] parts

    /// The parts for this pattern.
    member this.Parts = parts

    override this.Evaluate parameter =
        let builder = StringBuilder ()
        let rec evaluateParts parts =
            match parts with
            | (part : Pattern<'a, string>) :: parts ->
                match part.Evaluate parameter with
                | Some str -> 
                    builder.Append str |> ignore
                    evaluateParts parts
                | None -> None
            | [] -> Some (builder.ToString ())
        evaluateParts parts
    override this.Match target accum initial =
        let rec applyParts parts index accum initial  =
            match parts with
            | Known initial literal :: parts ->
                let hasLiteral = (literal.Length <= target.Length - index && String.substringEquals literal 0 target index literal.Length)
                if hasLiteral then applyParts parts (index + literal.Length) accum initial
                else accum
            | part :: Known initial suffix :: [] ->
                let partEnd = target.Length - suffix.Length
                let hasSuffix = (partEnd >= index && String.substringEquals suffix 0 target partEnd suffix.Length)
                if hasSuffix then part.Match (String.between target index partEnd) accum initial
                else accum
            | part :: Known initial delimiter :: parts ->
                match String.findOccurences target delimiter index with
                | [] -> accum
                | partEnd :: [] -> 
                    match part.Match (String.between target index partEnd) [] initial with
                    | [] -> accum
                    | state :: [] -> applyParts parts (partEnd + delimiter.Length) accum state
                    | states -> List.fold (applyParts parts (partEnd + delimiter.Length)) accum states
                | partEnds ->
                    let mutable accum = accum
                    for partEnd in partEnds do 
                        for parameter in part.Match (String.between target index partEnd) [] initial do
                            accum <- applyParts parts  (partEnd + delimiter.Length) accum parameter
                    accum
            | part :: [] -> part.Match (String.between target index target.Length) accum initial
            | part :: parts ->
                let mutable accum = accum
                for partEnd = index to target.Length do
                    for parameter in part.Match (String.between target index partEnd) [] initial do
                        accum <- applyParts parts partEnd accum parameter
                accum
            | [] when index = target.Length -> initial :: accum
            | [] -> accum
        applyParts parts 0 [] initial
    override this.Substitute map = Concat<'a>.Create (parts |> List.map (fun part -> part.Substitute map))

/// Contains functions for constructing and manipulating patterns.
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module Pattern =

    /// Creates a literal pattern.
    let literal value = Literal<'a, 'b> value :> Pattern<'a, 'b>

    /// Creates a string concatenation pattern.
    let concat parts = Concat<'a>.Create parts

    /// Substitutes the variables in the given pattern.
    let substitute map (pattern : Pattern<'a, 'b>) = pattern.Substitute map